### PR TITLE
Add support for allocations that exceed the PageRun limit

### PR DESCRIPTION
**Pull request by majetideepak**
This change enables an allocation to exceed the PageRun limit. We do this by splitting an allocation
across multiple PageRuns.
Non-contiguous allocations in the MallocAllocator use malloc to allocate and store the address
to later verify the allocation. Since an allocation can now span across multiple PageRuns, we need to
store the MachinePageCount for each allocation to identify allocation boundaries across the PageRuns.
Fix stats for `recordFree` in MallocAllocator nonContiguousFree.

### CODE REVIEW

File: velox/common/memory/MallocAllocator.cpp [hunk 1]
```diff
@@ -242,20 +242,28 @@ int64_t MallocAllocator::freeNonContiguous(Allocation& allocation) {
...
     }
-    stats_.recordFree(
-        std::min<int64_t>(
-            AllocationTraits::pageBytes(sizeClassSizes_.back()),
-            AllocationTraits::pageBytes(run.numPages())),
-        [&]() {
-          ::free(ptr); // NOLINT
-        });
+    numFreed += numPages;
+    stats_.recordFree(AllocationTraits::pageBytes(numPages), [&]() {
```
Comment (majetideepak) [inline velox/common/memory/MallocAllocator.cpp:256]: @xiaoxmeng This seems to have been incorrect previously.

File: velox/common/memory/MallocAllocator.h [hunk 2]
```diff
@@ -161,8 +161,12 @@ class MallocAllocator : public MemoryAllocator {
...
   std::mutex mallocsMutex_;
 
-  /// Tracks malloc'd pointers to detect bad frees.
-  std::unordered_set<void*> mallocs_;
+  /// Tracks malloc'd pointers and the corresponding MachinePageCount in
+  /// non-contiguous allocations to detect bad frees.
+  /// Since an allocation can span across multiple PageRuns, we need to store
+  /// the MachinePageCount for each allocation to identify allocation boundaries
+  /// across the PageRuns.
+  std::unordered_map<void*, MachinePageCount> mallocs_;
```
Comment (xiaoxmeng) [inline velox/common/memory/MallocAllocator.h:169]:
```
s/mallocs_/nonContiguousMallocs_/
```

File: velox/common/memory/MallocAllocator.cpp [hunk 3]
```diff
@@ -242,20 +242,28 @@ int64_t MallocAllocator::freeNonContiguous(Allocation& allocation) {
...
   for (int32_t i = 0; i < allocation.numRuns(); ++i) {
     Allocation::PageRun run = allocation.runAt(i);
-    numFreed += run.numPages();
     void* ptr = run.data();
+    MachinePageCount numPages;
     {
       std::lock_guard<std::mutex> l(mallocsMutex_);
-      const auto ret = mallocs_.erase(ptr);
-      VELOX_CHECK_EQ(ret, 1, "Bad free page pointer: {}", ptr);
+      const auto mapElement = mallocs_.find(ptr);
```
Comment (xiaoxmeng) [inline velox/common/memory/MallocAllocator.cpp:249]:
```
nit: s/mapElement/it/
```

File: velox/common/memory/MallocAllocator.cpp [hunk 4]
```diff
@@ -242,20 +242,28 @@ int64_t MallocAllocator::freeNonContiguous(Allocation& allocation) {
...
-        });
+    numFreed += numPages;
+    stats_.recordFree(AllocationTraits::pageBytes(numPages), [&]() {
+      ::free(ptr); // NOLINT
+    });
+    auto pagesInCurrentRun = run.numPages();
+    // It is possible that an allocation spans multiple PageRuns.
+    while (numPages > pagesInCurrentRun) {
+      numPages -= pagesInCurrentRun;
+      VELOX_CHECK_LT(++i, allocation.numRuns());
```
Comment (xiaoxmeng) [inline velox/common/memory/MallocAllocator.cpp:263]:
```
++i;
VELOX_CHECK_LT(i, allocation.numRuns());
```

File: velox/common/memory/MallocAllocator.cpp [hunk 5]
```diff
@@ -242,20 +242,28 @@ int64_t MallocAllocator::freeNonContiguous(Allocation& allocation) {
   MachinePageCount numFreed = 0;
   for (int32_t i = 0; i < allocation.numRuns(); ++i) {
     Allocation::PageRun run = allocation.runAt(i);
-    numFreed += run.numPages();
     void* ptr = run.data();
+    MachinePageCount numPages;
```
Comment (xiaoxmeng) [inline velox/common/memory/MallocAllocator.cpp:246]:
```
int64_t numPages;
```
So it can become negative to break out of the loop below? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@majetideepak LGTM % minors. Thanks!

**Comment by majetideepak** [comment]
@xiaoxmeng addressed your comments. Thanks!

**Review by xiaoxmeng — APPROVED** [review]

**Pull request closed without merging**